### PR TITLE
build(deps): use craft-application from main branch

### PR DESCRIPTION
**Pull request by lengau**
- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox run -m lint`?
- [ ] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

### CODE REVIEW

File: snapcraft/application.py [hunk 1]
```diff
@@ -22,21 +22,60 @@
...
+                )
+
+        return build_plan
+
 
 APP_METADATA = AppMetadata(
     name="snapcraft",
     summary="Package, distribute, and update snaps for Linux and IoT",
     ProjectClass=models.Project,
+    BuildPlannerClass=SnapcraftBuildPlanner,
```
Comment (sergiusens) [inline snapcraft/application.py:78]: feels like we need an AppSolver or similar now. Just a thought.

@cmatsuoka ?

**Review by mr-cal — APPROVED** [review]
LGTM

File: snapcraft/application.py [hunk 2]
```diff
@@ -22,21 +22,70 @@
...
 from snapcraft.commands import unimplemented
+from snapcraft.models import Architecture
+from snapcraft.providers import SNAPCRAFT_BASE_TO_PROVIDER_BASE
+from snapcraft.utils import get_effective_base, get_host_architecture
+
+
+class SnapcraftBuildPlanner(craft_application.models.BuildPlanner):
+    """A project model that creates build plans."""
+
+    def get_build_plan(self) -> List[BuildInfo]:
```
Comment (mr-cal) [inline snapcraft/application.py:46]: Should `snapcraft.models.project.Project.get_build_plan()` be removed as part of this PR?  Or is this an "add-before-remove" strategy?

**Review by mr-cal — APPROVED** [review]
Forgot to submit my comment

File: snapcraft/application.py (hunk 2 above)
Comment (lengau) [reply snapcraft/application.py:46]: Removing it caused some breakages - I think we need some upstream changes to fix that.

File: tests/spread/core24/patchelf/classic-python/task.yaml [hunk 3]
```diff
@@ -29,7 +29,8 @@ execute: |
   # Build the core24 snap
   unset SNAPCRAFT_BUILD_ENVIRONMENT
   # If this next line fails, delete it and replace it with the one following it
-  snapcraft --use-lxd 2>&1 | MATCH '"pack" command is not implemented for core24!'; exit 0
+  snapcraft --use-lxd 2>&1 | MATCH 'unrecognized arguments: --use-lxd'; exit 0
+  # snapcraft --use-lxd 2>&1 | MATCH '"pack" command is not implemented for core24!'; exit 0
```
Comment (sergiusens) [inline tests/spread/core24/patchelf/classic-python/task.yaml:33]:
```suggestion
```

File: snapcraft/application.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/application.py:46]: please create a task for this

Comment (lengau) [reply snapcraft/application.py:46]: https://github.com/canonical/craft-application/issues/208

**Review by sergiusens — APPROVED** [review]

**Pull request merged**